### PR TITLE
Add functions to support interleaved complex num

### PR DESCRIPTION
**Pull request by smu160**
@calebzulawski @Shnatsel 

A draft PR to address issue #22.

### CODE REVIEW

**Comment by smu160** [comment]
Hi @calebzulawski,

We now have the basics of what we'd need to support signals that are stored as an array of complex nums. Looking forward to hearing your thoughts!

Best,
Saveliy

File: Cargo.toml
Comment (Shnatsel) [inline Cargo.toml:16]: I think this should be an optional feature, maybe even a non-default one but enabled on docs.rs. The associated functions then need to be annotated as available with this feature only; there is a rustdoc feature for that.

File: src/lib.rs
Comment (Shnatsel) [inline src/lib.rs:192]: `chunks_exact_mut` is actually perfectly fine as long as you also handle the `.remainder()` separately. It's way faster because it's easy for the compiler to vectorize.

But I have a better idea: IIRC `zune-image` has deinterleaving as a simple, safe iterator that vectorizes into top notch code. Just reuse that, stick `multiversion` attribute on it, and it should get way way faster.

**Comment by calebzulawski** [comment]
Something like this looks good to me! Are you open to possibly using a more trait-oriented design in the future? I wouldn't want to hold up this functionality, but I think various input and output types could be combined into one generic function.

**Review by calebzulawski — APPROVED** [review]

**Comment by smu160** [comment]
Hi @calebzulawski,

Thank you for the review! I think your trait-oriented solution would be great. Just to clarify, do we want to wait on merging this PR? Or do we want this PR as an interim solution?

Best,
Saveliy

**Comment by calebzulawski** [comment]
I would probably merge it to have something to start from--just something to consider before a 1.0 release

**Comment by Shnatsel** [comment]
I've taken the liberty of rewriting `separate_re_im` and pushing to your branch.

I couldn't get it to vectorize [on godbolt](https://godbolt.org/z/E43nG83f6), but at least it should have the right public API, and we can always optimize it later.

**Comment by Shnatsel** [comment]
I tried a bunch more things and couldn't get it to autovectorize: https://godbolt.org/z/7Kd8v4vPG

We might need something like bytemuck here to view a `Complex` as a `&[T; 2]` or even better `&[Complex<T>]` as `&[T]` with double the length

**Comment by calebzulawski** [comment]
I think the approach I would take is iterate by chunks, bytemuck to `[T; N * 2]`, [Simd::deinterleave](https://doc.rust-lang.org/core/simd/struct.Simd.html#method.deinterleave), wrapped in `#[multiversion]`

**Comment by Shnatsel** [comment]
Yeah that sounds good to me

**Comment by calebzulawski** [comment]
I am working on a simd-complex crate--this seems like an important enough algorithm to include there as an AoSoA <-> AoS conversion

File: Cargo.toml
Comment (smu160) [reply Cargo.toml]: I made the following changes:

- `num-complex` is now a optional and non-default feature
- associated functions and tests are annotated as available with this feature only
- doc will not creates docs for all features (not sure if that's the desired outcome here?)
- update the CI pipeline to run tests for all features (since I had to annotate tests that use `Complex`)

**Comment by smu160** [comment]
@Shnatsel 

I have a rough version of `bytemuck` + [Simd deinterleave](https://doc.rust-lang.org/std/simd/struct.Simd.html#method.deinterleave), on [godbolt](https://godbolt.org/z/dPTsavhzG).

**Comment by Shnatsel** [comment]
I've taken a look at the Godbolt version. Wow, that is some aggressive loop unrolling! It looks good to me though.

One more thing to try: we need to benchmark if iterating over `vec![0; n]` is faster or slower than `Vec::with_capacity(n)` plus `Vec::extend_from_slice(deinterleaved.to_array())` in the main loop. The latter incurs slightly more instructions but avoids a memset to zero out both vectors.

File: src/lib.rs
Comment (Shnatsel) [inline src/lib.rs:235]: Why do we manually keep track of the index? I cannot see any places where it is actually used, is it an unused variable?

**Comment by smu160** [comment]
@Shnatsel I added the implementation using `bytemuck` + portable SIMD `deinterleave`. The solution in godbolt had a bounds check, so that was removed as well. You can see the difference in [godbolt](https://godbolt.org/z/xo54EvE1h).

File: src/lib.rs
Comment (smu160) [reply src/lib.rs:235]: It's used when taking care of the remainder

Comment (Shnatsel) [reply src/lib.rs:235]: Oh nevermind, I see it is used at the end. I am pretty sure we can infer it from the input length rather than keeping track of it in the hot loop. Something along the lines of `i = complex_f64.len() % CHUNK_SIZE` perhaps?

Comment (smu160) [reply src/lib.rs]: @Shnatsel Let me know if the checked-in change suffices for this

Comment (Shnatsel) [reply src/lib.rs]: Looks good to me performance-wise, I defer to tests to verify correctness

**Comment by smu160** [comment]
@Shnatsel 
> One more thing to try: we need to benchmark if iterating over vec![0; n] is faster or slower than Vec::with_capacity(n) plus Vec::extend_from_slice(deinterleaved.to_array()) in the main loop. The latter incurs slightly more instructions but avoids a memset to zero out both vectors.

I added a benchmark, using criterion, to test this idea. The benchmark is probably not leveraging criterion to its full potential, so let me know.

**Review by Shnatsel — APPROVED** [review]
Looks good, let's get this merged!

**Pull request merged**
